File: src/main/java/com/hoolix/processor/models/IntermediateEvent.java
package com.hoolix.processor.models;

import lombok.AllArgsConstructor;
import lombok.Getter;
import org.elasticsearch.action.index.IndexRequest;
import java.io.Serializable;
import java.util.Map;

/**
 * Created by peiyuchao on 2017/1/3.
 */
@AllArgsConstructor
@Getter
public class IntermediateEvent extends Event implements Serializable {
    private final Map<String, Object> payload;

    @Override
    public String getIndexName() {
        String token = payload.get("token").toString();
        String type = getType();
        Long uploadTimestamp = ((Long) payload.get("upload_timestamp"));
        Long eventTimestamp = ((Long) payload.getOrDefault("event_timestamp", null));
        // TODO 日期轮转
        if (eventTimestamp != null) {
            return token + Event.INDEX_NAME_SEPARATOR + type + eventTimestamp + uploadTimestamp;
        } else {
            return token + Event.INDEX_NAME_SEPARATOR + type + Event.INDEX_NAME_WILDCARD + uploadTimestamp;
        }
    }

    @Override
    public String getType() {
        return payload.get("type").toString();
    }

    @Override
    public IndexRequest toIndexRequest() {
        return new IndexRequest(getIndexName(), getType());
    }

    @Override
    public Map<String, Object> toPayload() {
        return payload;
    }
}

File: src/main/scala/com/hoolix/processor/models/IntermediateEvent.scala
package com.hoolix.processor.models

import org.elasticsearch.action.index.IndexRequest

/**
  * Created by peiyuchao on 2017/1/5.
  */
case class IntermediateEvent(
                              payload: collection.mutable.Map[String, Any]
                            ) extends Event {

  override def toPayload: collection.mutable.Map[String, Any] = payload

  override def indexName: String = {
    val token = payload.get("token") match {
      case Some(x: String) => x
      case _ => "_unknown_"
    }
//    val `type` = getType
//    val uploadTimestamp = payload.get("upload_timestamp").asInstanceOf[Some[Long]].get
//    val eventTimestamp = payload.getOrDefault("event_timestamp", null).asInstanceOf[Long]
    // TODO 日期轮转
//    if (eventTimestamp != null) token + Event.INDEX_NAME_SEPARATOR + `type` + eventTimestamp + uploadTimestamp
//    else token + Event.INDEX_NAME_SEPARATOR + `type` + Event.INDEX_NAME_WILDCARD + uploadTimestamp
    token + "." + indexType
  }

  override def indexType: String = payload.get("type") match {
    case Some(x: String) => x
    case _ => "_"
  }

  override def docId = ???
  override def toIndexRequest = new IndexRequest(indexName, indexType) // let ES generate a UUID for this
}
